pkg/controller/meta_pod_controller.go

```diff
@@ -0,0 +1,175 @@
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// Tutorial: https://grpc.io/docs/languages/go/quickstart/
+// Generate proto files via make proto
+
+// MetaPodController reconciles meta pods object.
+type MetaPodController struct {
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;update;patch
+
+// metaLeaderStatus sends a MetaMember request to a meta node at ip:port, determining if the node is a leader.
+func (mpc *MetaPodController) metaLeaderStatus(ctx context.Context, host string, port uint) string {
+	log := log.FromContext(ctx)
+
+	addr := fmt.Sprintf("%s:%v", host, port)
+
+	for i := 0; i < 5; i++ {
+		time.Sleep(time.Duration(i*10) * time.Millisecond)
+		conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			log.Error(err, "Unable to connect to meta pod. Retrying...")
+			continue
+		}
+		defer conn.Close()
+		c := pb.NewMetaMemberServiceClient(conn)
+
+		resp, err := c.Members(ctx, &pb.MembersRequest{})
+		if err != nil {
+			log.Error(err, "Sending MembersRequest failed")
+			continue
+		}
+
+		for _, member := range resp.Members {
+			if member.IsLeader && host == member.Address.Host && port == uint(member.Address.Port) {
+				return consts.MetaRoleLeader
+			}
+		}
+		return consts.MetaRoleFollower
+	}
+	return consts.MetaRoleUnknown
+}
+
+// getMetaPort returns the service port of the pod.
+func getMetaPort(pod *corev1.Pod) (uint, error) {
+	for _, container := range pod.Spec.Containers {
+		if container.Name == "meta" {
+			for _, containerPort := range container.Ports {
+				if containerPort.Name == "service" {
+					return uint(containerPort.ContainerPort), nil
+				}
+			}
+		}
+	}
+	return 0, fmt.Errorf("unable to retrieve the service port from pod %s", pod.ObjectMeta.Name)
+}
+
+// Reconcile handles the pods of the meta service. Will add the metaLeaderLabel to the pods.
+func (mpc *MetaPodController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	requeueInterval60s := time.Second * time.Duration(60)
+	defaultRequeueResult := ctrl.Result{RequeueAfter: requeueInterval60s}
+
+	// only reconcile when this is related to a meta pod
+	if !strings.Contains(req.Name, "meta") {
+		return defaultRequeueResult, nil
+	}
+
+	log := log.FromContext(ctx)
+
+	// get all meta pods
+	metaPods := &corev1.PodList{}
+	labels := labels.SelectorFromSet(labels.Set{"risingwave/component": "meta"})
+	listOptions := client.ListOptions{LabelSelector: labels}
+	if err := mpc.Client.List(context.Background(), metaPods, &listOptions); err != nil {
+		log.Error(err, "unable to fetch meta pods")
+		return ctrl.Result{Requeue: true}, err
+	}
```

Comment:
Why do we have to list the Pods? That makes the reconciliation busy because every change on any Pod will trigger a full update for all Pods. It doesn't make sense to me. I think here it should only update the label of the target Pod.

Reply:
IMHO reconciling all meta pods at once reduces the time we mislabel a pod during failover. 

Consider the following situation: 
```txt
time 1:
meta1 is a follower labeled as follower
meta2 is a follower labeled as follower
meta3 is a leader   labeled as leader
---
time 2: 
meta3/leader fails
meta2 takes over as new leader
meta3 is a New pod starts as follower
---
time 3:
meta2 is reconciled
meta1 is a follower labeled as follower
meta2 is a leader   labeled as leader
meta3 is a follower labeled as leader
```

Let me know if you agree with this. If not I will change it.

Reply:
Reconcile all Pods will be a sequential process, and then a timeout will block all others. But reconciling a Pod one time makes them parallel due to the nature of the framework. There's no way to avoid mislabelling entirely because the observation delay can not be eliminated. However, we can optimize the labels to eliminate the contradictions -- multiple leaders, by correcting the label of the old leader when a new leader is found.

I originally meant that the listing targeted all meta Pods in Kubernetes, and it wasn't a good pattern.

```go
labels.SelectorFromSet(labels.Set{"risingwave/component": "meta"})
```

Reply:
As discussed I will only update all pods that belong to the same tenant, IF we have a leader change.